pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/IntermediateSegment.java

```diff
@@ -224,6 +225,20 @@ public GenericRow getRecord(int docId, GenericRow reuse) {
     return reuse;
   }
 
+  @Override
+  public void getPrimaryKey(int docId, PrimaryKey reuse) {
+    int numPrimaryKeyColumns = _schema.getPrimaryKeyColumns().size();
+    Object[] values = new Object[numPrimaryKeyColumns];;
```

Comment:
We don't want to create new value buffer, use the one within the reuse primary key instead. Same for other places
```suggestion
    Object[] values = reuse.getValues();
```